lib/spack/spack/directives.py

```diff
@@ -932,6 +950,205 @@ def _execute_requires(pkg: Type[spack.package_base.PackageBase]):
+        return filtered
+
+    def handle_intersection(self, when, directive_entry, complement_versions, filtered):
+        new_when = when.copy()
+        new_when.versions = when.versions.intersection(complement_versions)
+        if new_when.versions != spack.version.VersionList():
+            self.add_to_filtered(filtered, new_when, directive_entry)
+
+    def iterate_directives(self, directives):
+        return directives
+
+
+class DropConflicts(DropDirectiveBase):
+    @property
+    def name(self):
+        return "conflicts"
+
+    def add_to_filtered(self, filtered, when, directive_entry):
+        filtered.setdefault(when, []).append(directive_entry)
+
+    def get_spec(self, directive_entry):
+        return directive_entry[0]
+
+
+class DropDependsOn(DropDirectiveBase):
+    @property
+    def name(self):
+        return "dependencies"
+
+    def add_to_filtered(self, filtered, when, directive_entry):
+        name, directive = directive_entry
+        filtered.setdefault(when, {})[name] = directive
+
+    def iterate_directives(self, directives):
+        return directives.items()
+
+    def get_spec(self, directive_entry):
+        return directive_entry[1].spec
+
+
+class DropRequires(DropConflicts):
+    @property
+    def name(self):
+        return "requirements"
+
+    def get_spec(self, directive_entry):
+        return directive_entry[0][0]
+
+
+@directive("conflicts")
+def drop_conflict(conflict_spec: SpecType, when: WhenType = None):
+    """Remove a conflict from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires to keep some of the conflicts in the parent package but delete others.
+
+    This code will not throw an error if the user inputs a conflict to delete
+    that does not exist.
+    """
+    return DropConflicts().remove(conflict_spec, when)
+
+
+@directive("dependencies")
+def drop_depends_on(spec: SpecType, when: WhenType = None):
+    """Remove a dependency from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires to keep some of the dependencies in the parent package but delete others.
+
+    This code will not throw an error if the user inputs a dependency to delete
+    that does not exist.
+    """
+    return DropDependsOn().remove(spec, when)
+
+
+@directive("requirements")
+def drop_requires(spec: SpecType, when: WhenType = None):
+    """Remove a dependency from a package.
+
+    This is typically used when inheriting from another package if the author
+    desires to keep some of the dependencies in the parent package but delete others.
+
+    This code will not throw an error if the user inputs a dependency to delete
+    that does not exist.
+    """
+    return DropRequires().remove(spec, when)
+
+
+@directive("versions")
+def drop_version(ver: Union[str, int]):
+    """Try to remove a specific version from a package."""
```

Comment:
Can this be made into `drop_versions()`, and can it accept a `VersionRange` or a `VersionList`? 

Ideally this would use satisfies to figure out what to drop so that I can drop ranges of versions, e.g., given:

```python
class Gcc(Package):
    version("14.2.0")
    version("9.5.0")
    version("8.5.0")
    version("8.3.0")
    version("5.4.0")
    version("5.2.0")
    version("4.9.3")
    version("4.9.2")
    version("4.9.1")
```

You could write:

```
drop_versions("=4.9.2,5.4.0:8,14")
```

and get:

```python
class MyrGcc(Package):
    version("9.5.0")
    version("5.2.0")
    version("4.9.3")
    version("4.9.1")
```

Can that work?

Reply:
Yes, I'm modifying it right now to do that.  I'm currently adding support for DropPatch and DropVersion but have had to generalize the code a bit because of their different function signatures.  Not a big deal but I probably won't finish until next week.  The good news is that the generalization should make it apply to all of the directives.

Reply:
I have `drop_patch` working now and will do `drop_version(s)` next week.  I've had some time to think about your request for drop_versions.  I'm still open to doing the `drop_versions` modification but have realized that some of the directives like `drop_patch` accept a lot of arguments so they probably won't be compatible with the version range idea.  Even `drop_version` will have a similar issue when it supports the `when` argument.  I've just been doing `[drop_version(ver) for ver in ["1.3", "1.1"]]` to achieve something similar.  Anyways, I'll update everyone when this is ready for another review.